### ISSUE TITLE
Permission denied error on accessing repository cache on Windows

### ISSUE THREAD

**Issue by sgowroji** [bazelbuild/bazel#21823]
CI: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/3716#018e3609-27ef-44b6-a03c-5aba25e8d3f2

[https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/3716#018e3609-27df-40d8-ba7c-7b356e1eb89e](https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/3716#018e3609-27df-40d8-ba7c-7b356e1eb89e)

Platform: Windows, MacOS

Logs:

```
ERROR: [0mAn error occurred during the fetch of repository '_main~rules_dotnet_dev_nuget_packages_extension~nuget.microsoft.netcore.app.runtime.win-x64.v6.0.8':
Exception: C:/b/cache/repos/v1/content_addressable/sha256/38f308a5ea5a54387770ab1efcadfed3ec556fe22d9dd3d9d1d14c0171810837/file (Permission denied)
ERROR: [0m<builtin>: fetching nuget_archive rule //:_main~rules_dotnet_dev_nuget_packages_extension~nuget.microsoft.netcore.app.runtime.win-x64.v6.0.8: Traceback (most recent call last):
Exception: C:/b/cache/repos/v1/content_addressable/sha256/38f308a5ea5a54387770ab1efcadfed3ec556fe22d9dd3d9d1d14c0171810837/file (Permission denied)
ERROR: [0mno such package '@@_main~rules_dotnet_dev_nuget_packages_extension~nuget.microsoft.netcore.app.runtime.win-x64.v6.0.8//': java.io.IOException: C:/b/cache/repos/v1/content_addressable/sha256/38f308a5ea5a54387770ab1efcadfed3ec556fe22d9dd3d9d1d14c0171810837/file (Permission denied)
ERROR: [0mC:/b/qxmqlmv5/external/_main~rules_dotnet_dev_nuget_packages_extension~paket.rules_dotnet_dev_nuget_packages/microsoft.netcore.app.runtime.win-x64/6.0.8/BUILD.bazel:5:15: @@_main~rules_dotnet_dev_nuget_packages_extension~paket.rules_dotnet_dev_nuget_packages//microsoft.netcore.app.runtime.win-x64/6.0.8:6.0.8 depends on @@_main~rules_dotnet_dev_nuget_packages_extension~nuget.microsoft.netcore.app.runtime.win-x64.v6.0.8//:microsoft.netcore.app.runtime.win-x64.6.0.8.nupkg in repository @@_main~rules_dotnet_dev_nuget_packages_extension~nuget.microsoft.netcore.app.runtime.win-x64.v6.0.8 which failed to fetch. no such package '@@_main~rules_dotnet_dev_nuget_packages_extension~nuget.microsoft.netcore.app.runtime.win-x64.v6.0.8//': java.io.IOException: C:/b/cache/repos/v1/content_addressable/sha256/38f308a5ea5a54387770ab1efcadfed3ec556fe22d9dd3d9d1d14c0171810837/file (Permission denied)
ERROR: [0mAnalysis of target '//dotnet/private/tests/use_as_tool/fsharp:run_self_contained' failed; build aborted: Analysis failed
ERROR: [0mBuild did NOT complete successfully
FAILED:[0m 
FAILED:[0m 
Error: The command exited with status 1[0m
error: exit status 1
```

Culprit: bazelbuild/bazel@af6e090

bazelbuild/bazel@8eade04

bazelbuild/bazel@63be83f

Steps:

```
 git clone https://github.com/bazelbuild/rules_dotnet
git reset 2af3c4d68510d9631d71e61caea919c28636b218  --hard
export USE_BAZEL_VERSION=98f29db0ac47b4b3c75bccdcfda20ffd50521597
bazel build //... 
```

CC Greenteam @Wyverald

**Comment by meteorcloudy** [comment]
I actually don't quite understand what exact situation triggered this error. Is it because multiple `put` actions trying to write the cache at the same time?

I saw there was a comment `TODO(jingwen): Implement file locking for concurrent cache accesses.`, maybe we could do so? /cc @jin

**Comment by fmeum** [comment]
> I actually don't quite understand what exact situation triggered this error. Is it because multiple `put` actions trying to write the cache at the same time?

It's ultimately hard to say for certain, but there can be multiple `put` actions that try to create the cache file while there is already a `get` accessing the file put there by an earlier `put`. That's why the atomicity of the `put` isn't by itself sufficient to solve this.

> I saw there was a comment `TODO(jingwen): Implement file locking for concurrent cache accesses.`, maybe we could do so? /cc @jin 

This would be an alternative. Since the cached file is only ever in one of two states (non-existent or the correct content), I don't think that file locking would be much easier though. I will look into what it would take though.

**Comment by meteorcloudy** [comment]
Ah,, I see, this is really tricky. Thanks for looking into it!

**Comment by doak** [comment]
Was that only an issue on a Windows environment or is there any chance that a very similar issue happens on Linux as well (and is not fixed by #21861)?
I don't want to hijack this issue. I just get very similar cache issues since switching to to v7.0.0 and it is still reproducibly for v7.3.1. Any pointers a very much appreciated.
